### PR TITLE
[Signature discovery] Fetch a signed image manifest for parsing

### PR DESCRIPTION
**Pull request by yawangwang**
Adding a client library to fetch a image manifest using a tag-based discovery mechanism.

### CODE REVIEW

**Comment by yawangwang** [comment]
cc: @haydentherapper to review this PR for any insightful comments.

**Review by haydentherapper — APPROVED** [review]
This looks good to me. This is following the steps outlined in https://github.com/sigstore/cosign/blob/79d4184cef8326738166b430a0f08339840b2da6/specs/SIGNATURE_SPEC.md?plain=1#L151-L169.

File: launcher/signature-discovery/client/client.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+	c := &Client{
+		cdClient:      cdClient,
+		OriginalImage: originalImage,
+		RemoteOpts:    opts,
+	}
+	return c
+}
+
+// FetchSignedImageManifest fetches a signed image manifest using a tag-based discovery mechanism.
+func (c *Client) FetchSignedImageManifest(ctx context.Context, targetRepository string) (v1.Manifest, error) {
```
Comment (jkl73) [inline launcher/signature-discovery/client/client.go:33]: Is this the function going to be called periodically?

**Review by jkl73 — APPROVED** [review]

File: launcher/signature-discovery/client/client.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+// New creates a new client.
+func New(cdClient *containerd.Client, originalImage containerd.Image, opts ...containerd.RemoteOpt) *Client {
+	c := &Client{
+		cdClient:      cdClient,
+		OriginalImage: originalImage,
+		RemoteOpts:    opts,
+	}
+	return c
```
[the remark this answers is not in the document]
Comment (yawangwang) [reply launcher/signature-discovery/client/client.go]: Done.

File: launcher/signature-discovery/client/client.go (hunk 1 above)
Comment (yawangwang) [reply launcher/signature-discovery/client/client.go:32]: > Is this the function going to be called periodically?

Yes, this function will be ultimately called inside `refreshToken`. This change will be reflected on the following PR.

File: launcher/signature-discovery/client/client.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+		return v1.Manifest{}, err
+	}
+	return getManifest(ctx, image)
+}
+
+func (c *Client) pullTargetImage(ctx context.Context, targetRepository string) (containerd.Image, error) {
+	targetImageRef := fmt.Sprint(targetRepository, ":", formatSigTag(c.OriginalImage))
+	image, err := c.cdClient.Pull(ctx, targetImageRef, c.RemoteOpts...)
+	if err != nil {
+		return nil, fmt.Errorf("[signature-discovery]: cannot pull the image [%s] from taregetRepository [%s]: %w", targetImageRef, targetRepository, err)
```
Comment (alexmwu) [inline launcher/signature-discovery/client/client.go:44]: We generally don't have prefixes for components in our logs. To keep our logging consistent, let's do something like: `"cannot pull signature object '%s' from target repository '%s': %w"`

File: launcher/signature-discovery/client/client_test.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/defaults"
+	"github.com/containerd/containerd/namespaces"
+)
+
+func pullPublicImage(ctx context.Context, imageRef string, cdClient *containerd.Client) (containerd.Image, error) {
```
Comment (alexmwu) [inline launcher/signature-discovery/client/client_test.go:13]: is there a way we can avoid actually pulling down from gcr?

**Review by alexmwu — APPROVED** [review]
Be sure to squash the commits if you merge.

File: launcher/signature-discovery/client/client_test.go (hunk 4 above)
Comment (yawangwang) [reply launcher/signature-discovery/client/client_test.go]: Removed this function and replaced the actual image with an image descriptor in `Client` struct.

File: launcher/signature-discovery/client/client.go
Comment (alexmwu) [inline launcher/signature-discovery/client/client.go:1]: The subdirectory should be named either signaturediscovery or signature_discovery. Do we need a nested client directory? What else will go in signaturediscovery?

File: launcher/signature-discovery/client/client.go [hunk 5]
```diff
@@ -14,35 +14,42 @@ const signatureTagSuffix = "sig"
...
+		return v1.Manifest{}, err
 	}
 	return getManifest(ctx, image)
 }
 
+func (c *Client) pullTargetImage(ctx context.Context, targetRepository string) (containerd.Image, error) {
+	targetImageRef := fmt.Sprint(targetRepository, ":", formatSigTag(c.OriginalImageDesc))
+	image, err := c.cdClient.Pull(ctx, targetImageRef, c.RemoteOpts...)
+	if err != nil {
+		return nil, fmt.Errorf("cannot pull the signature object [%s] from tareget repository [%s]: %w", targetImageRef, targetRepository, err)
```
Comment (alexmwu) [inline launcher/signature-discovery/client/client.go:44]: very nit: tareget -> target

**Review by alexmwu — APPROVED** [review]

File: launcher/signature-discovery/client/client.go
Comment (alexmwu) [reply launcher/signature-discovery/client/client.go:1]: let's also put it in an internal directory. we want to make it usable by other packages in the library, but NOT export it to users of the library. so `launcher/internal/signaturediscovery/discovery.go`, `launcher/internal/signaturediscovery/client.go` or something like that

Comment (yawangwang) [reply launcher/signature-discovery/client/client.go:1]: Done. A new library for parsing signed container image object will be added to `internal` directory in the upcoming PR.

File: launcher/signature-discovery/client/client.go (hunk 5 above)
Comment (yawangwang) [reply launcher/signature-discovery/client/client.go]: > very nit: tareget -> target

Done

**Pull request merged**